android/app/src/main/java/org/ZingoLabs/Zingo/BackgroundSyncWorker.kt

```diff
@@ -0,0 +1,125 @@
+package org.ZingoLabs.Zingo
+
+import android.content.Context
+import androidx.work.Worker
+import androidx.work.WorkerParameters
+import android.util.Log
+import java.io.File
+import java.util.*
+import org.json.JSONObject
+import java.io.InputStream
+import java.nio.charset.StandardCharsets
+import com.facebook.react.bridge.ReactApplicationContext
+
+class BackgroundSyncWorker(context: Context, workerParams: WorkerParameters) : Worker(context, workerParams) {
+
+    override fun doWork(): Result {
+        val reactContext = ReactApplicationContext(MainApplication.getAppContext())
+        val rpcModule = RPCModule(reactContext)
+
+        Log.w("SCHEDULED_TASK_RUN", "Task running")
+
+        // checking if the wallet file exists
+        val exists: Boolean = walletExists()
+
+        if (exists) {
+            RustFFI.initlogging()
+            
+            // check the Server, because the task can run without the App.
+            val balance = RustFFI.execute("balance", "")
+            Log.w("SCHEDULED_TASK_RUN", "Testing if server is active: $balance")
+            if (balance.lowercase().startsWith("error")) {
+                // this means this task is running with the App closed
+                loadWalletFile(rpcModule)
+            } else {
+                // this means the App is open,
+                // stop syncing first, just in case.
+                stopSyncingProcess()
+            }
+
+            // interrupt sync to false, just in case it is true.
+            val resp = RustFFI.execute("interrupt_sync_after_batch", "false")
+            Log.w("SCHEDULED_TASK_RUN", "Not interrupting sync: $resp")
+
+            // the task is running here blocking this execution until this process finished:
+            // 1. finished the syncing.
+
+            Log.w("SCHEDULED_TASK_RUN", "sync BEGIN")
+            val resp2 = RustFFI.execute("sync", "")
+            Log.w("SCHEDULED_TASK_RUN", "sync END: $resp2")
+
+        } else {
+            Log.w("SCHEDULED_TASK_RUN", "No exists wallet file END")
+            return Result.failure()
+
+        }
+
+        // save the wallet file with the new data from the sync process
+        rpcModule.saveWallet()
+        Log.w("SCHEDULED_TASK_RUN", "wallet file SAVED")
+
+        // save the background JSON file
+        val timeStamp = Date().time / 1000
+        val timeStampStr = timeStamp.toString()
+        val jsonBackground = "{\"batches\": \"0\", \"date\": \"$timeStampStr\"}"
+        rpcModule.saveBackgroundFile(jsonBackground)
+        Log.w("SCHEDULED_TASK_RUN", "background json file SAVED")
+
+        return Result.success()
+    }
+
+    private fun loadWalletFile(rpcModule: RPCModule) {
+        // I have to init from wallet file in order to do the sync
+        // and I need to read the settings.json to find the server & chain type
+        val file: InputStream = MainApplication.getAppContext()?.openFileInput("settings.json")!!
```

Comment:
When I work with a closable component, I prefer using `.use{} ` to auto-close it and avoid leaks when anything goes wrong. See https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.io/use.html

Reply:
Thanks, I did like you suggested.